k9mail/src/main/java/com/fsck/k9/ui/messageview/MessageViewFragment.java

```diff
@@ -408,7 +408,14 @@ public void onCopy() {
     }
 
     public void onArchive() {
-        onRefile(mAccount.getArchiveFolderName());
+        if (mMessage != null) {
+            if (mAccount.isUseFolderStructureWhenArchive()) {
+                String archiveFolder = mAccount.getArchiveFolderName() + "." + mMessage.getFolder().getName();
+                onRefile(archiveFolder);
+            } else {
+                onRefile(mAccount.getArchiveFolderName());
```

Comment:
This snippet is similar to the changes you made to `MessageListFragment`, could you extract the code into a method, something like 

```
String calculateArchiveFolderName(Account account, Folder folder) {
    if (account.isUseFolderStructureWhenArchive()) {
        return account.getArchiveFolderName() + "." + folder.getName();
    } else {
        return account.getArchiveFolderName();
    }
}
```

and call it from both places?